src/trio/_util.py

```diff
@@ -353,3 +357,61 @@ def wraps(  # type: ignore[explicit-any]
 
 else:
     from functools import wraps  # noqa: F401  # this is re-exported
+
+
+def _raise(exc: BaseException) -> NoReturn:
+    """This helper allows re-raising an exception without __context__ being set."""
+    # cause does not need special handling, we simply avoid using `raise .. from ..`
+    __tracebackhide__ = True
+    context = exc.__context__
+    try:
+        raise exc
+    finally:
+        exc.__context__ = context
+        del exc, context
+
+
+def raise_single_exception_from_group(
+    eg: BaseExceptionGroup[BaseException],
+) -> NoReturn:
+    """This function takes an exception group that is assumed to have at most
+    one non-cancelled exception, which it reraises as a standalone exception.
+
+    If a :exc:`KeyboardInterrupt` is encountered, a new KeyboardInterrupt is immediately
+    raised with the entire group as cause.
+
+    If the group only contains :exc:`Cancelled` it reraises the first one encountered.
+
+    It will retain context and cause of the contained exception, and entirely discard
+    the cause/context of the group(s).
+
+    If multiple non-cancelled exceptions are encountered, it raises
+    :exc:`AssertionError`.
+    """
+    cancelled_exceptions = []
+    noncancelled_exceptions = []
+
+    # subgroup/split retains excgroup structure, so we need to manually traverse
+    def _parse_excg(e: BaseException) -> None:
+        if isinstance(e, KeyboardInterrupt):
+            # immediately bail out
+            raise KeyboardInterrupt from eg
+
+        if isinstance(e, trio.Cancelled):
+            cancelled_exceptions.append(e)
```

Comment:
I haven't looked at the PR which introduced this to trio-websocket but I'm a bit surprised by this... It's not necessarily wrong since then the next await point will add in `Cancelled`, but it feels like this isn't exactly necessary. The only reason a `Cancelled` would show up here is that some `CancelScope` further up the stack is cancelled, in which case any other exception is a failure in cancel-cleanup code and should probably be acknowledged.

I get the case for at least compressing `BaseExceptionGroup(..., [Cancelled._create(), Cancelled._create()])` though. Is there some other case I'm missing? Or is this maybe just a simpler implementation? I don't care much about this, it just feels weird to special case `Cancelled` here.

Reply:
the reason this special-cases `Cancelled` is to compress `BaseExceptionGroup(..., [ArbitraryError, Cancelled])` -> `ArbitraryError` without having to panic and raise `AssertionError`.
Does that answer the question? If not then I'm not quite following

Reply:
Hm, not quite. I don't know how to really explain myself, it just feels weird that we special case Cancelled given that that will only happen if something outside of the code that's raising is cancelled, eg:

```
with trio.CancelScope() as cancel_scope, this_as_a_context_manager:
  async with trio.open_nursery() as nursery:
    nursery.start_soon(trio.sleep, 5)
    await trio.sleep(1)
    cancel_scope.cancel()
    raise ValueError("bleh")
```

Like in the above case it's not obvious to me that we should only keep the ValueError.

Reply:
yeah I mean all alternatives suck one way or another. But unless we want to raise `AssertionError` I think it's much better to drop the cancellation (since it can trigger again) than to drop the other exception.
But maybe cancel+exception is not expected to come up in any remotely normal usage (with e.g. #3197) and everything should panic if it does. idk I'm not sure what the usage patterns would look like

Reply:
Alright as long as you've considered it. It doesn't seem that bad, just slightly... weird, I suppose?